autogluon/utils/tabular/ml/trainer/abstract_trainer.py

```diff
@@ -867,225 +574,636 @@ def refit_single_full(self, X=None, y=None, X_val=None, y_val=None, models=None)
+            The model's name is appended to self.models_level[stack_name][level]
+            The model's base_models (if it has any) must all be a lower level than the model.
+        """
+        if model.val_score is not None and stack_name != self.distill_stackname:  # TODO: may want to avoid hard-coding logic into specific stack names
+            logger.log(20, '\t' + str(round(model.val_score, 4)) + '\t = Validation ' + self.eval_metric.name + ' score')
+        if model.fit_time is not None:
+            logger.log(20, '\t' + str(round(model.fit_time, 2)) + 's' + '\t = Training runtime')
+        if model.predict_time is not None:
+            logger.log(20, '\t' + str(round(model.predict_time, 2)) + 's' + '\t = Validation runtime')
+        # TODO: Add to HPO
+        # TODO/FIXME: what if it isn't valid?
+        if model.is_valid():
+            if isinstance(model, BaggedEnsembleModel):
+                type_inner = model._child_type
+            else:
+                type_inner = type(model)
+            self.model_graph.add_node(
+                model.name,
+                fit_time=model.fit_time,
+                predict_time=model.predict_time,
+                val_score=model.val_score,
+                path=model.path,
+                type=type(model),  # Outer type, can be BaggedEnsemble, StackEnsemble (Type that is able to load the model)
+                type_inner=type_inner,  # Inner type, if Ensemble then it is the type of the inner model (May not be able to load with this type)
+                can_infer=model.can_infer(),
+            )
+            if isinstance(model, StackerEnsembleModel):
+                # TODO: raise exception if any base models are of equal or higher level?
+                # TODO: raise exception if no base models and level != 0?
+                for stack_column_prefix in model.stack_column_prefix_lst:
+                    base_model_name = model.stack_column_prefix_to_model_map[stack_column_prefix]
+                    self.model_graph.add_edge(base_model_name, model.name)
+            stack_loc = self.models_level[stack_name]  # TODO: Consider removing, have _train_multi handle this
+            if model.name not in stack_loc[level]:
+                stack_loc[level].append(model.name)
+        if self.low_memory:
+            del model
+
+    # TODO: Split this to avoid confusion, HPO should go elsewhere?
+    def _train_single_full(self, X_train, y_train, model: AbstractModel, X_val=None, y_val=None, feature_prune=False,
+                           hyperparameter_tune=False, stack_name='core', k_fold=None, k_fold_start=0, k_fold_end=None, n_repeats=None, n_repeat_start=0, level=0, time_limit=None) -> List[str]:
+        """
+        Trains a model, with the potential to train multiple models via hyperparameter tuning and feature pruning.
+        Returns a list of successfully trained and saved model names.
```

Comment:
maybe clarify these will be accessible in this Trainer

Reply:
added